### PR TITLE
Electra: exclude empty requests in requests list

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

 Other

**What does this PR do? Why is it needed?**

we need to add in the first byte as the request type to the ssz serialized list of requests and avoid appending the list if it's empty.

**Which issues(s) does this PR fix?**

Fixes # https://github.com/ethereum/execution-apis/pull/599

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: proto/engine/v1/electra.go [hunk 1]
```diff
@@ -19,35 +19,45 @@ var (
...
-	requests.Withdrawals = wrs
-
-	// consolidation requests
-	crs, err := unmarshalItems(ebe.ExecutionRequests[2], crSize, func() *ConsolidationRequest { return &ConsolidationRequest{} })
-	if err != nil {
-		return nil, err
+	var prevTypeNum uint8
+	for i := range ebe.ExecutionRequests {
+		requestType := ebe.ExecutionRequests[i][0]
+		if prevTypeNum > requestType {
```
Comment (james-prysm) [inline proto/engine/v1/electra.go:33]: would like an opinion on this check

File: beacon-chain/execution/engine_client_test.go [hunk 2]
```diff
@@ -1769,7 +1769,7 @@ func fixturesStruct() *payloadFixtures {
 			Proofs:      []hexutil.Bytes{[]byte("proof1"), []byte("proof2")},
 			Blobs:       []hexutil.Bytes{{'a'}, {'b'}},
 		},
-		ExecutionRequests: []hexutil.Bytes{depositRequestBytes, withdrawalRequestBytes, consolidationRequestBytes},
+		ExecutionRequests: []hexutil.Bytes{append([]byte{0}, depositRequestBytes...), append([]byte{1}, withdrawalRequestBytes...), append([]byte{2}, consolidationRequestBytes...)},
```
Comment (prestonvanloon) [inline beacon-chain/execution/engine_client_test.go:1772]:
```suggestion
		ExecutionRequests: []hexutil.Bytes{append([]byte{depositRequestType}, depositRequestBytes...), append([]byte{withdrawalRequestType}, withdrawalRequestBytes...), append([]byte{consolidationRequestType}, consolidationRequestBytes...)},
```

Replace the magic numbers with their constants

File: proto/engine/v1/electra.go [hunk 3]
```diff
@@ -19,35 +19,45 @@ var (
...
-	}
-	requests.Withdrawals = wrs
-
-	// consolidation requests
-	crs, err := unmarshalItems(ebe.ExecutionRequests[2], crSize, func() *ConsolidationRequest { return &ConsolidationRequest{} })
-	if err != nil {
-		return nil, err
+	var prevTypeNum uint8
+	for i := range ebe.ExecutionRequests {
+		requestType := ebe.ExecutionRequests[i][0]
```
Comment (prestonvanloon) [inline proto/engine/v1/electra.go:32]: This method could panic if the request is empty slice when you index `[1:]` later. 
Consider a creating a method that handles this like:

```go
func decodeExecutionRequest(req []byte) (typ int, data []byte, err error) {
  if len(req) < 1 {
    return -1, nil, errors.New("invalid execution request, length less than 1")
  }
  return int(req[0]), req[1:], nil
}
```

Then you can use it safely in your for-range loop

```go
reqType, reqData, err := decodeExecutionRequest(ebe.ExecutionRequests[i])
```

File: proto/engine/v1/electra.go [hunk 4]
```diff
@@ -57,21 +67,38 @@ func EncodeExecutionRequests(requests *ExecutionRequests) ([]hexutil.Bytes, erro
...
-	wrBytes, err := marshalItems(requests.Withdrawals)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to marshal withdrawal requests")
+	// request types MUST be in sorted order starting from 0
+	if len(requests.Deposits) > 0 {
+		drBytes, err := marshalItems(requests.Deposits)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal deposit requests")
+		}
+		requestData := []byte{0}
```
Comment (prestonvanloon) [inline proto/engine/v1/electra.go:78]:
```suggestion
		requestData := []byte{depositRequestType}
```

File: proto/engine/v1/electra.go (hunk 3 above)
Comment (james-prysm) [reply proto/engine/v1/electra.go:32]: good catch, there's another case we should probably catch too

File: proto/engine/v1/electra.go [hunk 5]
```diff
@@ -19,59 +20,112 @@ var (
...
+		requestType, requestListInSSZBytes, err := decodeExecutionRequest(ebe.ExecutionRequests[i])
+		if err != nil {
+			return nil, err
+		}
+		if prevTypeNum > requestType {
+			return nil, errors.New("invalid execution request type order, requests should be in sorted order")
+		}
+		prevTypeNum = requestType
+		switch requestType {
+		case DepositRequestType:
```
Comment (potuz) [inline proto/engine/v1/electra.go:42]: As before can we remove the cases from here and treat them in separate functions?

**Review by terencechain — APPROVED** [review]

**Pull request merged**